aten/src/ATen/native/TensorShape.cpp

```diff
@@ -1315,81 +1315,349 @@ Tensor index_select_sparse(const Tensor& self, int64_t dim, const Tensor& index)
+      else {
+        std::tie(index_intersect, dim_indices_intersect, n_intersect, res_nnz)
+          = compute_index_intersections_and_nnz(
+              sel_indices_unique, sel_indices_counts, n_unique_sel_indices,
+              dim_indices_unique, dim_indices_counts, n_unique_dim_indices
+            );
       }
-      for (const auto j : c10::irange(nnz)) {
-        int64_t jdx = cpu_dim_indices_ptr[j];
-        if (idx == jdx) {
-          new_nnz++;
-          iindices.push_back(i);
-          zindices.push_back(j);
+
+      const auto compute_offsets = [](const Tensor& counts, int64_t len) {
+        const auto narrowed_counts = counts.narrow(-1, 0, len);
+        return narrowed_counts.cumsum(/*dim=*/0).sub_(narrowed_counts);
+      };
+
+      const auto dim_sort_indices_offsets = compute_offsets(dim_indices_counts, n_unique_dim_indices);
+      const auto sel_sort_indices_offsets = compute_offsets(sel_indices_counts, n_unique_sel_indices);
+
+      auto selected_dim_indices = at::empty({res_nnz}, dim_indices_unique.options());
+      auto res_dim_indices = at::empty({res_nnz}, sel_indices_unique.options());
+      {
+        auto* ptr_selected_dim_indices = selected_dim_indices.data_ptr<int64_t>();
+        auto* ptr_res_dim_indices = res_dim_indices.data_ptr<int64_t>();
+
+        const auto* ptr_dim_indices_intersect = dim_indices_intersect.data_ptr<int64_t>();
+        const auto* ptr_sel_indices_intersect = index_intersect.data_ptr<int64_t>();
+
+        const auto* ptr_dim_indices_counts = dim_indices_counts.data_ptr<int64_t>();
+        const auto* ptr_sel_indices_counts = sel_indices_counts.data_ptr<int64_t>();
+
+        const auto* ptr_dim_sort_indices = dim_sort_indices.data_ptr<int64_t>();
+        const auto* ptr_sel_sort_indices = sel_sort_indices.data_ptr<int64_t>();
+
+        const auto* ptr_dim_indices_offsets = dim_sort_indices_offsets.data_ptr<int64_t>();
+        const auto* ptr_sel_indices_offsets = sel_sort_indices_offsets.data_ptr<int64_t>();
+
+        for (const auto ii : c10::irange(n_intersect)) {
+          const auto i_idx = *ptr_dim_indices_intersect++;
+          const auto j_idx = *ptr_sel_indices_intersect++;
+
+          const auto i_idx_count = ptr_dim_indices_counts[i_idx];
+          const auto j_idx_count = ptr_sel_indices_counts[j_idx];
+
+          const auto i_idx_offset = ptr_dim_indices_offsets[i_idx];
+          const auto j_idx_offset = ptr_sel_indices_offsets[j_idx];
+
+          const auto* src_dim_sort_indices = ptr_dim_sort_indices + i_idx_offset;
+          const auto* src_sel_sort_indices = ptr_sel_sort_indices + j_idx_offset;
+
+          const auto copy_chunk_len = i_idx_count * j_idx_count;
+          for (const auto chunk_elem_idx : c10::irange(copy_chunk_len)) {
+            *ptr_selected_dim_indices++ = *(src_dim_sort_indices + (chunk_elem_idx % i_idx_count));
+            *ptr_res_dim_indices++ = *(src_sel_sort_indices + (chunk_elem_idx % j_idx_count));
+          }
```

Comment:
We can trivially parallelize either the outer or the inner loop. Maybe it is possible to efficiently parallelize them jointly...